src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java

```diff
@@ -0,0 +1,204 @@
+     * @param addAssetActionBuilder         function that takes an asset draft and an asset position to build an
+     *                                      AddAsset action of the type T.
+     * @param <T>                           the type of the resource the asset update actions are built for.
+     * @return a list of asset update actions on the resource of type T if the list of assets is not identical.
+     *         Otherwise, if the assets are identical, an empty list is returned.
+     * @throws BuildUpdateActionException in case there are asset drafts with duplicate keys.
+     */
+    @Nonnull
+    public static <T> List<UpdateAction<T>> buildAssetsUpdateActions(
+        @Nonnull final List<Asset> oldAssets,
+        @Nullable final List<AssetDraft> newAssetDrafts,
+        @Nonnull final BiFunction<Asset, AssetDraft, List<UpdateAction<T>>> assetActionsBuilder,
+        @Nonnull final Function<String, UpdateAction<T>> removeAssetActionBuilder,
+        @Nonnull final Function<List<String>, UpdateAction<T>> changeAssetOrderActionBuilder,
+        @Nonnull final BiFunction<AssetDraft, Integer, UpdateAction<T>> addAssetActionBuilder)
+        throws BuildUpdateActionException {
+
+        if (newAssetDrafts != null) {
+            return buildAssetsUpdateActionsWithNewAssetDrafts(
+                oldAssets,
+                newAssetDrafts,
+                assetActionsBuilder,
+                removeAssetActionBuilder,
+                changeAssetOrderActionBuilder,
+                addAssetActionBuilder);
+        } else {
+            return oldAssets.stream()
+                            .map(oldAsset -> removeAssetActionBuilder.apply(oldAsset.getKey()))
+                            .collect(Collectors.toList());
+        }
+    }
+
+    @Nonnull
+    private static <T> List<UpdateAction<T>> buildAssetsUpdateActionsWithNewAssetDrafts(
+        @Nonnull final List<Asset> oldAssets,
+        @Nonnull final List<AssetDraft> newAssetDrafts,
+        @Nonnull final BiFunction<Asset, AssetDraft, List<UpdateAction<T>>> assetActionsBuilder,
+        @Nonnull final Function<String, UpdateAction<T>> removeAssetActionBuilder,
+        @Nonnull final Function<List<String>, UpdateAction<T>> changeAssetOrderActionBuilder,
+        @Nonnull final BiFunction<AssetDraft, Integer, UpdateAction<T>> addAssetActionBuilder)
+        throws BuildUpdateActionException {
+
+        // Asset list that represents the state of the old variant assets after each applied update action.
+        final List<Asset> intermediateOldAssets = new ArrayList<>(oldAssets);
+
+        final Map<String, Asset> oldAssetsKeyMap = oldAssets.stream().collect(toMap(Asset::getKey, asset -> asset));
+
+        Map<String, AssetDraft> newAssetDraftsKeyMap;
+        try {
+            newAssetDraftsKeyMap = newAssetDrafts
+                .stream().collect(toMap(AssetDraft::getKey, assetDraft -> assetDraft));
+        } catch (final IllegalStateException exception) {
```

Comment:
Looks mis-concept and overhead: you verify keys duplicate catching `IllegalStateException`, but:
  - it is an overhead from performance perspective (exception handling creates a lot of logic under the hood)
  - `IllegalStateException` might mean not only keys duplicate, but also other state exceptions.

Better approach is to make custom iteration with duplicate keys matching, e.g. `Map#containsKey()`

Reply:
09bbc4c